### PR TITLE
Quasi-uniform grid support and coordinates origin at center

### PR DESCRIPTION
**Pull request by Noe-LV**
This PR creates a new grid policy called QuasiUniformGrid that lets the user set a constant mesh resolution in each space direction. It takes as arguments dx, dy, and dz that correspond to these constant spacings.

Importantly, this PR changes the order of operations during constraints resolution. All grid policies are first resolved as Rectilinear grids, using only the volume and grid policy specifications. Then placement constraints are resolved with a Rectilinear grid. This is consistent with the choice of keeping apart user-friendly grid policy classes such as UniformGrid, QuasiUniformGrid, possibly AutoGrid as in tidy3d, and an internal RectilinearGrid representation for the algorithms.

It also partly adresses #358 and tries to keep real-space coordinates consistent  with the representation choice of origin-at-the-center coordinates. However, RectilinearGrid coordinates are still computed as positive integers (e.g. for a cubic volume with length of 100µm and 1µm resolution, real coordinates are (-50e-6, 50e-6), but RectilinearGrid coordinates are (0, 100)). This needs to be updated.

Tests were updated for the most part. Some still use origin-at-lower-corner coordinates.

## Summary by CodeRabbit

* **New Features**
  * Added `QuasiUniformGrid` support with independent per-axis spacing, validation, and CFL-based timestep computation.
* **Improvements**
  * Updated the simulation domain to a center-based coordinate convention, improving resolved-grid snapping and geometry calculations.
  * Enhanced object placement and constraint solving by resolving and using the final grid geometry earlier.
  * Adjusted GDS coordinate alignment and voxelization to match the center-based convention.
* **Tests**
  * Updated unit/integration tests for the new centering, snapping, and expected coordinate/bounds behavior.

### CODE REVIEW

**Comment by ymahlau** [comment]
Nice @Noe-LV, many thanks for the PR! I will take a closer look at the code later today or tomorrow :)

File: src/fdtdx/core/grid.py [hunk 1]
```diff
@@ -67,21 +77,28 @@ def slice_extent(
...
+        a center-relative basis: ``coord`` is interpreted relative to
+        ``self.center[axis]`` and the returned index is a center-relative edge
+        offset. Use :meth:`RectilinearGrid.coord_to_index` on the resolved grid
+        when you need absolute edge indices.
+        """
+        origin_offset = coord - self.center[axis]
+        scaled = origin_offset / self.spacing
         if snap == "nearest":
-            return round(scaled)
+            return int(jnp.round(scaled))
```
Comment (ymahlau) [inline src/fdtdx/core/grid.py]: The "jnp" here instead of "np" would prevent using the function under jax.jit. Please revert

**Comment by ymahlau** [comment]
@Noe-LV apart from the comment above, I found nothing obvious at first glance. That being said it is a bit difficult to give a proper review when multiple things (new grid, bug fixes, grid definition, ...) are mixed in a single PR. In the future please address everything in smaller separate PRs. 

Of course all unit tests should run through before I can accept the PR

**Comment by ymahlau** [comment]
@coderabbitai review

**Comment by ymahlau** [comment]
also please merge the latest changes into this branch, there were some recent updates to docstring tests

**Comment by Noe-LV** [comment]
Thanks @ymahlau for the review! I've haven't had much experience yet in open-source contributing. Things got a little out of hand when trying to keep the changes small. I'll be careful next time. 

I fixed the last docstrings issues and merged the last changes from the main branch. I think it is ready to merge.
